### PR TITLE
Multisite support

### LINKED ISSUE #37: Support multisite setups
When running `drush dl devel` from /backdrop/sites/[sitename] I get the following message:

> Success:  Project devel downloaded to /backdrop/modules/devel.

It should download to /backdrop/sites/[sitename]/modules instead.

----

PR @BWPanda https://github.com/backdrop-contrib/drush/pull/143

### PR DESCRIPTION
**Pull request by ghost**
Fixes https://github.com/backdrop-contrib/drush/issues/37.

### CODE REVIEW

**Comment by serundeputy** [comment]
@BWPanda the code looks good, but I'm failing at spinning up a multisite; so I haven't been able to test locally yet :(

**Comment by ghost** [comment]
@serundeputy Can I help you with this somehow?

**Comment by serundeputy** [comment]
@BWPanda there are two possible ways to help me with this:

First way
----

1. Spin up a multisite and a non-multisite (normal backdrop).
  * Manually test that most/all commands still work in both (particularly the ones changed here `dl` and `up`)
  * Document what you did here and state your results

2. Spin up a D7 multisite and D7 regular
  * Manually test that most/all commands still work in both (particularly the ones changed here `dl` and `up`)
  * Document what you did here and state your results

(sometimes updates to these commands break functionality in D7 and we don't want to do that)

Second way 
----

You can teach me to fish, i.e. how to set up multisite in Backdrop, but I'm allergic to this one. 👍 🐼 👍

**Comment by ghost** [comment]
@serundeputy I vote option 2 :grin: Have sent you an email with instructions.

**Comment by serundeputy** [comment]
@BWPanda it seems I can't find the time to multisite ... so if you want to move this forward I think you'll have to do option `First way`.

>First way
Spin up a multisite and a non-multisite (normal backdrop).
Manually test that most/all commands still work in both (particularly the ones changed here dl and up)
Document what you did here and state your results
Spin up a D7 multisite and D7 regular
Manually test that most/all commands still work in both (particularly the ones changed here dl and up)
Document what you did here and state your results
(sometimes updates to these commands break functionality in D7 and we don't want to do that)

**Comment by ghost** [comment]
@serundeputy Ok, long time coming, but here we go...

## Backdrop

### Setup

1. Install Drush 8 and Backdrop Drush on server
1. Apply [this patch](https://patch-diff.githubusercontent.com/raw/backdrop-contrib/drush/pull/143.patch) to Backdrop Drush
1. Download and extract [Backdrop](https://github.com/backdrop/backdrop/releases/download/1.13.0/backdrop.zip) on server
1. Copy 'backdrop' directory to 'backdrop_multisite' (create two separate installations)

#### Normal

1. Setup virtual host pointing to `backdrop`
1. Create DB & user
1. Add entry to local hosts file
1. Install Backdrop (install.php)
1. Access site (make sure it works)

#### Multisite

1. Setup virtual host pointing to `backdrop_multisite`
1. Create DB & user
1. Add entry to local hosts file
1. Add the following to `sites/sites.php`: `$sites['[DOMAIN]'] = '[DIR]';`
1. Create directories: `sites/[DIR]`, `sites/[DIR]/files`, `sites/[DIR]/modules`, `sites/[DIR]/themes`, `sites/[DIR]/layouts`
1. Copy `settings.php` into `sites/[DIR]`
1. Install Backdrop (install.php)
1. Access site (make sure it works, and that `sites/[DIR]/files` was populated, not `files` in root directory)

### Testing

These tests needs to be run from the directory where the `settings.php` file resides (e.g. root for the normal installation, and `sites/[DIR]` for the multisite installation).

In each installation:
1. Run `drush dl simplify` - confirm the module downloads to the correct place
1. Run `drush dl borg` - confirm the theme downloads to the correct place
1. Run `drush dl senjo` - confirm the layout downloads to the correct place
1. Download and unzip [devel 1.6.0](https://github.com/backdrop-contrib/devel/releases/download/1.x-1.6.0/devel.zip) to the `modules` directory
1. Run `drush up devel` - confirm devel is updated in the correct place

### Results

All projects downloaded to the correct folder for the installation they were run from. Devel was updated in the correct place as well. In the multisite installation, all project folders and the `files` folder in the root directory were (correctly) empty (save for a README file).

## Drupal 7

### Setup

1. Install Drush 8 and Backdrop Drush on server
1. Apply [this patch](https://patch-diff.githubusercontent.com/raw/backdrop-contrib/drush/pull/143.patch) to Backdrop Drush
1. Download and extract [Drupal](https://ftp.drupal.org/files/projects/drupal-7.67.tar.gz) on server
1. Rename 'drupal-7.67' to 'drupal' and copy 'drupal' to 'drupal_multisite' (create two separate installations)

#### Normal

1. Setup virtual host pointing to `drupal`
1. Create DB & user
1. Add entry to local hosts file
1. Install Backdrop (install.php)
1. Access site (make sure it works)

#### Multisite

1. Setup virtual host pointing to `drupal_multisite`
1. Create DB & user
1. Add entry to local hosts file
1. Rename `sites/example.sites.php` to `sites/sites.php` and add the following to it: `$sites['[DOMAIN]'] = '[DIR]';`
1. Create directories: `sites/[DIR]`, `sites/[DIR]/files`, `sites/[DIR]/modules`, `sites/[DIR]/themes`
1. Copy `sites/default/default.settings.php` to `sites/[DIR]/settings.php`
1. Install Backdrop (install.php)
1. Access site (make sure it works, and that `sites/[DIR]/files` was populated, not `sites/default/files`)

### Testing

These tests needs to be run from the directory where the `settings.php` file resides (e.g. `sites/default` for the normal installation, and `sites/[DIR]` for the multisite installation).

In each installation:
1. Run `drush dl simplify` - confirm the module downloads to the correct place
1. Run `drush dl zen` - confirm the theme downloads to the correct place
1. Download and unzip [devel 1.6](https://ftp.drupal.org/files/projects/devel-7.x-1.6.zip) to the `modules` directory, then run `drush en devel`
1. Run `drush up devel` - confirm devel is updated in the correct place

### Results

All projects downloaded to the correct folder for the installation they were run from. Devel was updated in the correct place as well.

**Comment by ghost** [comment]
TLDR: Everything works as expected :smiley:

But would love if you (or others) can confirm that too.

File: commands/pm/backdrop_pm.drush.inc [hunk 1]
```diff
@@ -209,19 +209,26 @@ function backdrop_pm_get_tags($url) {
  */
 function backdrop_pm_get_path($tags) {
   $type = (string) $tags->type;
+  if (!empty(drush_get_context('DRUSH_BACKDROP_SITE_ROOT'))) {
```
Comment (drupol) [inline commands/pm/backdrop_pm.drush.inc:212]:
```suggestion
  $site_root = BACKDROP_ROOT;
  if (drush_get_context('DRUSH_BACKDROP_SITE_ROOT', NULL) !== NULL) {
```

File: commands/pm/backdrop_pm.drush.inc [hunk 2]
```diff
@@ -370,11 +377,18 @@ function backdrop_command_pm_update() {
       }
     }
     elseif (!empty($project)) {
-      if (file_exists("$backdrop_root/modules/contrib")) {
-        $module_path = "$backdrop_root/modules/contrib";
+      if (!empty(drush_get_context('DRUSH_BACKDROP_SITE_ROOT'))) {
```
Comment (drupol) [inline commands/pm/backdrop_pm.drush.inc:380]:
```suggestion
      if (drush_get_context('DRUSH_BACKDROP_SITE_ROOT', NULL) === NULL) {
```

File: commands/pm/backdrop_pm.drush.inc [hunk 3]
```diff
@@ -370,11 +377,18 @@ function backdrop_command_pm_update() {
...
+      else {
+        $site_root = $backdrop_root;
+      }
+
+      if (file_exists("$site_root/modules/contrib")) {
+        $module_path = "$site_root/modules/contrib";
       }
       else {
-        $module_path = "$backdrop_root/modules";
+        $module_path = "$site_root/modules";
```
Comment (drupol) [inline commands/pm/backdrop_pm.drush.inc:391]:
```suggestion

```

File: commands/pm/backdrop_pm.drush.inc [hunk 4]
```diff
@@ -370,11 +377,18 @@ function backdrop_command_pm_update() {
...
-      if (file_exists("$backdrop_root/modules/contrib")) {
-        $module_path = "$backdrop_root/modules/contrib";
+      if (!empty(drush_get_context('DRUSH_BACKDROP_SITE_ROOT'))) {
+        $site_root = realpath($backdrop_root . '/' . drush_get_context('DRUSH_BACKDROP_SITE_ROOT'));
+      }
+      else {
+        $site_root = $backdrop_root;
+      }
+
+      if (file_exists("$site_root/modules/contrib")) {
```
Comment (drupol) [inline commands/pm/backdrop_pm.drush.inc:387]:
```suggestion
      $module_path = $site_root . '/modules';
      if (file_exists($site_root . '/contrib')) {
```

File: commands/pm/backdrop_pm.drush.inc [hunk 5]
```diff
@@ -370,11 +377,18 @@ function backdrop_command_pm_update() {
...
+        $module_path = "$site_root/modules/contrib";
```
Comment (drupol) [inline commands/pm/backdrop_pm.drush.inc:388]:
```suggestion
        $module_path = $site_root . '/contrib';
```

File: commands/pm/backdrop_pm.drush.inc [hunk 6]
```diff
@@ -370,11 +377,18 @@ function backdrop_command_pm_update() {
...
       }
       else {
```
Comment (drupol) [inline commands/pm/backdrop_pm.drush.inc:390]:
```suggestion
      
```

File: commands/pm/backdrop_pm.drush.inc [hunk 7]
```diff
@@ -370,11 +377,18 @@ function backdrop_command_pm_update() {
...
-        $module_path = "$backdrop_root/modules";
+        $module_path = "$site_root/modules";
       }
```
Comment (drupol) [inline commands/pm/backdrop_pm.drush.inc:392]:
```suggestion

```

File: commands/pm/backdrop_pm.drush.inc [hunk 8]
```diff
@@ -209,19 +209,26 @@ function backdrop_pm_get_tags($url) {
...
 function backdrop_pm_get_path($tags) {
   $type = (string) $tags->type;
+  if (!empty(drush_get_context('DRUSH_BACKDROP_SITE_ROOT'))) {
+    $site_root = realpath(BACKDROP_ROOT . '/' . drush_get_context('DRUSH_BACKDROP_SITE_ROOT'));
+  }
+  else {
+    $site_root = BACKDROP_ROOT;
+  }
+
   if ($type == 'project_module') {
```
Comment (drupol) [inline commands/pm/backdrop_pm.drush.inc:219]:
```suggestion
   if ($type === 'project_module') {
```

File: commands/pm/backdrop_pm.drush.inc [hunk 9]
```diff
@@ -209,19 +209,26 @@ function backdrop_pm_get_tags($url) {
  */
 function backdrop_pm_get_path($tags) {
   $type = (string) $tags->type;
+  if (!empty(drush_get_context('DRUSH_BACKDROP_SITE_ROOT'))) {
+    $site_root = realpath(BACKDROP_ROOT . '/' . drush_get_context('DRUSH_BACKDROP_SITE_ROOT'));
+  }
+  else {
```
Comment (drupol) [inline commands/pm/backdrop_pm.drush.inc:215]:
```suggestion

```

File: commands/pm/backdrop_pm.drush.inc [hunk 10]
```diff
@@ -209,19 +209,26 @@ function backdrop_pm_get_tags($url) {
...
+    $site_root = BACKDROP_ROOT;
```
Comment (drupol) [inline commands/pm/backdrop_pm.drush.inc:216]:
```suggestion

```

File: commands/pm/backdrop_pm.drush.inc [hunk 11]
```diff
@@ -209,19 +209,26 @@ function backdrop_pm_get_tags($url) {
...
+  }
```
Comment (drupol) [inline commands/pm/backdrop_pm.drush.inc:217]:
```suggestion

```

**Review by drupol — CHANGES_REQUESTED** [review]
Here's a quick review.
